tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -0,0 +1,580 @@
+package org.unicode.cldr.util.personname;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeSet;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.ibm.icu.util.ULocale;
+
+/**
+ * Rough sketch for now
+ * TODO Make classes/methods private that don't need to be public
+ * TODO Check for invalid parameters
+ */
+
+public class PersonNameFormatter {
+    public enum Field {
+        prefix,
+        given,
+        given2,
+        surname,
+        surname2,
+        suffix;
+    }
+
+    public enum Length {
+        long_name,
+        medium_name,
+        narrow_name,
+        monogram,
+        monogram_narrow
+    }
+
+    public enum Style {
+        formal,
+        informal
+    }
+
+    public enum Usage {
+        referring,
+        addressing,
+        sorting
+    }
+
+    public enum Order {
+        surname_first,
+        given_first
+    }
+
+
+    public enum Modifier {
+        initial,
+        all_caps,
+        informal,
+        core
+    }
+
+    public static final Splitter SPLIT_SPACE = Splitter.on(' ').trimResults();
+    public static final Splitter SPLIT_DASH = Splitter.on('-').trimResults();
+    public static final Splitter SPLIT_EQUALS = Splitter.on('=').trimResults();
+    public static final Joiner JOIN_SPACE = Joiner.on(' ');
+    public static final Joiner JOIN_DASH = Joiner.on('-');
+
+
+    /**
+     * A Field and its modifiers, corresponding to a string form like {given-initial}.
+     * Immutable
+     */
+    public static class ModifiedField {
```

Comment:
It is simpler to have this as an object, since it is used in a few other places.

Reply:
That's what I'm challenging though. I don't think should be used in "few other places". :) Once `SimpleNameObject` is out, `NamePatternElement` doesn't need this at all.

Reply:
Right, but it is useful in the implementation of NameObject subclasses.

Reply:
Again, remember that this will not be public API; this will be within the CLDR tooling. It also does not need to be tuned for performance. However, it can reflect on the algorithm in the spec; if we think that it cannot be optimized we may choose a different algorithm

Reply:
I'm not sure I'm on board with this stance. The spec not only defines the schema for personal name formats, but also talks about "default implementation". If this work is not the "default implementation", then what is? I think, if there's no real "default implementation" ("public API"), then treating this work as one is in fact valuable, because it allows us to both validate decisions that we've made in the spec and provide a good starting point for adopters of the standard. The latter is important, because whether we like it or not, someone _will_ treat this as a "public API" and base their implementation on it.

Reply:
Ah, I understand now. Here's what we normally plan for:

* In CLDR we have an API that we can use for testing and examples. The timeline is data submission (in this case, May 11 for CLDR v42). That gets used to gather the initial data. It should fully implement the spec, but does not have to be optimized. While someone could dig the code out of the CLDR sources, normally we haven't had a problem with that.
* ICU will develop a production-level API, first as a tech-preview (in this case, for release in October). It doesn't have to follow the CLDR internal API. 
* Once the ICU API is stable enough, we cut over to for internal use in CLDR. (The ICU API does has to have enough 'hooks' (where we can supply prospective CLDR data) that we can eventually use for gathering data in CLDR.) That could be in CLDR v43 or v44.

Does this address your concern?

Reply:
Yes, thanks.